src/execution/execute.ts

```diff
@@ -605,12 +612,32 @@ function handleFieldError(
     throw error;
   }
 
+  // Do not modify errors list if a response position for this error has already been nulled.
+  const errorPositions = exeContext.errorPositions;
+  if (hasNulledPosition(errorPositions, path)) {
+    return null;
+  }
+  errorPositions.add(path);
+
   // Otherwise, error protection is applied, logging the error and resolving
   // a null value for this field if one is encountered.
   exeContext.errors.push(error);
   return null;
 }
 
+function hasNulledPosition(
+  errorPositions: Set<Path | undefined>,
+  path: Path | undefined,
+): boolean {
+  if (errorPositions.has(path)) {
+    return true;
+  }
+  if (path === undefined) {
+    return false;
+  }
+  return hasNulledPosition(errorPositions, path.prev);
+}
```

Comment:
Recursion can be expensive in JS, let's use a loop instead.

```suggestion
function hasNulledPosition(
  errorPositions: Set<Path | undefined>,
  startPath: Path | undefined,
): boolean {
  let path = startPath;
  while (path !== undefined) {
    if (errorPositions.has(path)) {
      return true;
    }
    path = path.prev;
  }
  return errorPositions.has(undefined);
}
```